metadata-ingestion/tests/integration/tableau/test_tableau_ingest.py

```diff
@@ -1365,3 +1366,87 @@ def test_permission_mode_switched_error(pytestconfig, tmp_path, mock_datahub_gra
                 "Turn on your derived permissions. See for details "
                 "https://community.tableau.com/s/question/0D54T00000QnjHbSAJ/how-to-fix-the-permissionsmodeswitched-error"
             )
+
+
+@freeze_time(FROZEN_TIME)
+def test_extract_project_hierarchy():
+    context = PipelineContext(run_id="0", pipeline_name="test_tableau")
+
+    config_dict = config_source_default.copy()
+
+    del config_dict["stateful_ingestion"]
+    del config_dict["projects"]
+
+    config_dict["project_pattern"] = {
+        "allow": ["project1", "project4"],
+        "deny": ["project2"],
+    }
+
+    config_dict["extract_project_hierarchy"] = True
+
+    config = TableauConfig.parse_obj(config_dict)
+
+    site_source = TableauSiteSource(
+        config=config,
+        ctx=context,
+        platform="tableau",
+        site=SiteItem(name="Site 1", content_url="site1"),
+        site_id="site1",
+        report=TableauSourceReport(),
+        server=Server("https://test-tableau-server.com"),
+    )
+
+    all_project_map: Dict[str, TableauProject] = {
+        "p1": TableauProject(
+            id="1",
+            name="project1",
+            path=[],
+            parent_id=None,
+            parent_name=None,
+            description=None,
+        ),
+        "p2": TableauProject(
+            id="2",
+            name="project2",
+            path=[],
+            parent_id="1",
+            parent_name="project1",
+            description=None,
+        ),
+        "p3": TableauProject(
+            id="3",
+            name="project3",
+            path=[],
+            parent_id="1",
+            parent_name="project1",
+            description=None,
+        ),
+        "p4": TableauProject(
+            id="4",
+            name="project4",
+            path=[],
+            parent_id=None,
+            parent_name=None,
+            description=None,
+        ),
+    }
+
+    site_source._init_tableau_project_registry(all_project_map)
+
+    # "project1", "project4" are included because of `allow` pattern
+    # "project3" is included because it is not explicitly denied in `deny` and `extract_project_hierarchy` is enabled
+    assert ["project1", "project4", "project3"] == [
+        project.name for project in site_source.tableau_project_registry.values()
+    ]
+
+    # reset the registry to test extract_project_hierarchy=false
+    site_source.tableau_project_registry = {}
+
+    config.extract_project_hierarchy = False
```

Comment:
these ideally should've been two separate unit tests, not smushed into a single test

Reply:
I kept it the same to ensure that anyone referencing it can clearly see the distinction between true and false.

Reply:
then you should use a parameterized test